Support EDM-style training in DreamBooth LoRA SDXL script

sayakpaul opened the pull request:
Command example:

```bash
CUDA_VISIBLE_DEVICES=1 accelerate launch train_dreambooth_lora_sdxl.py \
  --pretrained_model_name_or_path="playgroundai/playground-v2.5-1024px-aesthetic"  \
  --instance_data_dir="dog" \
  --output_dir="dog-playground-lora" \
  --mixed_precision="fp16" \
  --instance_prompt="a photo of sks dog" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --learning_rate=1e-4 \
  --use_8bit_adam \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=500 \
  --validation_prompt="A photo of sks dog in a bucket" \
  --validation_epochs=25 \
  --seed="0" \
  --push_to_hub
```

WandB: https://wandb.ai/sayakpaul/dreambooth-lora-playground/runs/sxe4bavp

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1764 @@
+                        return_dict=False,
+                    )[0]
+
+                model_pred = model_pred * (-sigmas) + noisy_model_input
```
Same comment as above, this formula is used when applying euler to DDPM models. For EDM it's

```py
c_skip = sigma_data**2 / (sigma**2 + sigma_data**2)
c_out = sigma * sigma_data / (sigma**2 + sigma_data**2) ** 0.5

model_pred = c_skip * noisy_model_input + c_out * model_pred
```

Let's use the `precondition_outputs` method here.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_7126). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1764 @@
+                    )[0]
+
+                model_pred = model_pred * (-sigmas) + noisy_model_input
+                weighing = sigmas**-2.0
```
For later: this could be made configurable, as there are multiple weighing alternatives. In EDM they use 
https://github.com/NVlabs/edm/blob/main/training/loss.py#L75

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1764 @@
+                weighing = sigmas**-2.0
+
+                # Get the target for loss depending on the prediction type
+                if noise_scheduler.config.prediction_type == "epsilon":
+                    target = noise
+                elif noise_scheduler.config.prediction_type == "v_prediction":
+                    target = noise_scheduler.get_velocity(model_input, noise, timesteps)
+                else:
+                    raise ValueError(f"Unknown prediction type {noise_scheduler.config.prediction_type}")
```
For EDM, the `target` should be the `latents` (i.e `model_input` here).

sayakpaul commented:
@patil-suraj could you give this another look?

sayakpaul commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1765 @@
+
+    # Move unet, vae and text_encoder to device and cast to weight_dtype
+    unet.to(accelerator.device, dtype=weight_dtype)
+    vae.to(accelerator.device, dtype=torch.float32)
```
Important to keep the VAE in fp32, otherwise it leads to numerical problems.

patil-suraj commented:
Can't seem to find anything else, will also try to run the script and see what's going on.

patil-suraj commented:
Got a good run: https://wandb.ai/psuraj/dreambooth-lora-playground/runs/j34izml0?workspace=user-psuraj (still going on)

What fixed it:
- Load `fp32` variant of the `vae` .
- don't use `autocast` during generation
- disable loss weighing.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1765 @@
+    # run inference
+    generator = torch.Generator(device=accelerator.device).manual_seed(args.seed) if args.seed else None
+
+    with torch.cuda.amp.autocast():
```
Let's remove this. This is causing overflows for some reason.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1765 @@
+        args.pretrained_model_name_or_path,
+        subfolder="vae",
+        revision=args.revision,
+        variant=args.variant,
```
```suggestion
```

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1765 @@
+                    )[0]
+
+                model_pred = noise_scheduler.precondition_outputs(noisy_model_input, model_pred, sigmas)
+                weighing = sigmas**-2.0
```
This causes `NaN` loss, let's remove for now
```suggestion
```

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1765 @@
+
+                    # Compute prior loss
+                    prior_loss = torch.mean(
+                        (weighing.float() * (model_pred_prior.float() - target_prior.float()) ** 2).reshape(
```
```suggestion
                        ((model_pred_prior.float() - target_prior.float()) ** 2).reshape(
```

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_playground.py:
```diff
@@ -0,0 +1,1765 @@
+                    prior_loss = prior_loss.mean()
+
+                loss = torch.mean(
+                    (weighing.float() * (model_pred.float() - target.float()) ** 2).reshape(target.shape[0], -1), 1
```
```suggestion
                    ((model_pred.float() - target.float()) ** 2).reshape(target.shape[0], -1), 1
```

patil-suraj commented:
Started a new run here https://wandb.ai/psuraj/dreambooth-lora-playground/runs/ef2qkmre

sayakpaul commented:
@patil-suraj ready for a review. Feel free to test the script too :) 

@pcuenca feel free to give this a review as well.

pcuenca commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -119,27 +136,35 @@ def save_model_card(
+    if "playgroundai" in base_model:
+        tags.extend(["playground", "playground-diffusers"])
+    else:
+        tags.extend(["stable-diffusion-xl", "stable-diffusion-xl-diffusers"])
```
Should these also be applicable, seeing that the playground model is essentially derivative?

pcuenca commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1034,8 +1083,12 @@ def main(args):
         vae_path,
         subfolder="vae" if args.pretrained_vae_model_name_or_path is None else None,
         revision=args.revision,
-        variant=args.variant,
+        variant=args.variant if "playgroundai" not in args.pretrained_model_name_or_path else None,
```
This looks a bit strange to me. An alternative would be to just recommend users to use the main variant for this model (they can still use mixed-precision training).

patil-suraj replied:
+1

sayakpaul replied:
Done in c8ed8af8. LMK.

pcuenca commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1034,8 +1083,12 @@ def main(args):
-        variant=args.variant,
+        variant=args.variant if "playgroundai" not in args.pretrained_model_name_or_path else None,
     )
+    if "playgroundai" in args.pretrained_model_name_or_path:
+        latents_mean = torch.tensor(vae.config.latents_mean).view(1, 4, 1, 1)
+        latents_std = torch.tensor(vae.config.latents_std).view(1, 4, 1, 1)
```
Can't this be derived from the configuration? If so, we can set a var accordingly and maybe try to avoid to use `"playgroundai"` in many places.

patil-suraj replied:
+1

sayakpaul replied:
Done in c8ed8af8. LMK.

pcuenca reviewed (approved):
Looks good in general. I'd maybe try to avoid hardcoded references to the string `"playgroundai"` to make decisions, if possible.

sayakpaul commented on examples/dreambooth/README_sdxl.md:
```diff
@@ -206,3 +206,41 @@ You can explore the results from a couple of our internal experiments by checkin
+
+It's now possible to perform EDM-style training as proposed in [Elucidating the Design Space of Diffusion-Based Generative Models](https://arxiv.org/abs/2206.00364). 
+
+For the SDXL model, simple set:
```
There's a test that you can check but I haven't done a full-blown training run.

patil-suraj replied:
For LoRA it might not work, but can def be fine-tuned with EDM.

sayakpaul replied:
@patil-suraj elaborate?

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -159,23 +184,29 @@ def log_validation(
 
     # run inference
     generator = torch.Generator(device=accelerator.device).manual_seed(args.seed) if args.seed else None
+    # Currently the context determination is a bit hand-wavy. We can improve it in the future if there's a better
+    # way to condition it. Reference: https://github.com/huggingface/diffusers/pull/7126#issuecomment-1968523051
+    inference_ctx = (
+        contextlib.nullcontext() if "playgroundai" in args.pretrained_model_name_or_path else torch.cuda.amp.autocast()
+    )
```
Very nice!

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1018,7 +1058,16 @@ def main(args):
     # Load scheduler and models
-    noise_scheduler = DDPMScheduler.from_pretrained(args.pretrained_model_name_or_path, subfolder="scheduler")
+    scheduler_type = determine_scheduler_type(args.pretrained_model_name_or_path, args.revision)
+    if scheduler_type == "EDMDPMSolverMultistepScheduler" or args.do_edm_style_training:
```
Here, instead of checking for a specific scheduler, we could check if the name has a prefix of `EDM`, that would be more general.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1018,7 +1058,16 @@ def main(args):
+    scheduler_type = determine_scheduler_type(args.pretrained_model_name_or_path, args.revision)
+    if scheduler_type == "EDMDPMSolverMultistepScheduler" or args.do_edm_style_training:
+        args.do_edm_style_training = True
+        noise_scheduler = EDMDPMSolverMultistepScheduler.from_pretrained(
```
We should use `EDMEuler` here. For training, we should only recommend DDPM and Euler/EDMEuler. Rest are specifically designed for inference sampling.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1512,22 +1582,37 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
-                if args.pretrained_vae_model_name_or_path is None:
-                    model_input = model_input.to(weight_dtype)
+
+                if "playgroundai" not in args.pretrained_model_name_or_path:
```
same comment as above. We should check of the `mean` and `std` args exist and not `None` instead of checking for `playgroundai`

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1512,22 +1582,37 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
+                        0, noise_scheduler.config.num_train_timesteps, (bsz,), device=model_input.device
+                    )
+                    timesteps = timesteps.long()
+                else:
+                    indices = torch.randint(0, noise_scheduler.config.num_train_timesteps, (bsz,))
+                    timesteps = noise_scheduler.timesteps[indices].to(device=model_input.device)
```
```suggestion
                else:
                    # in EDM formulation, the model is conditioned on the pre-conditioned noise levels instead of discrete timesteps, so here we sample indices to get the noise levels from `scheduler.timesteps`
                    indices = torch.randint(0, noise_scheduler.config.num_train_timesteps, (bsz,))
                    timesteps = noise_scheduler.timesteps[indices].to(device=model_input.device)
```

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1512,22 +1582,37 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
                 # Add noise to the model input according to the noise magnitude at each timestep
                 # (this is the forward diffusion process)
                 noisy_model_input = noise_scheduler.add_noise(model_input, noise, timesteps)
+                if args.do_edm_style_training:
+                    sigmas = get_sigmas(timesteps, len(noisy_model_input.shape), noisy_model_input.dtype)
+                    inp_noisy_latents = noise_scheduler.precondition_inputs(noisy_model_input, sigmas)
```
Let's add a comment here, explaining the pre-conditioning, could be also cool to link the formula in paper.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1570,16 +1655,23 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
                     )[0]
 
+                if args.do_edm_style_training:
+                    model_pred = noise_scheduler.precondition_outputs(noisy_model_input, model_pred, sigmas)
```
same comment as above.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1570,16 +1655,23 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
                 # Get the target for loss depending on the prediction type
                 if noise_scheduler.config.prediction_type == "epsilon":
-                    target = noise
+                    target = model_input if args.do_edm_style_training else noise
```
We should do this for `v-predition` as well, in EDM, the target is always the actual input as always denoise the image in `precondition_outputs`

patil-suraj replied:
That gives the velocity, but when using EDM, whether we are predicting `epsilon` or `v`, it is always converted to back to the sample, hence the target should be `model_input`. Loss is always computed against the actual image and predicted image and not `v` or `eps` . Does that make sense ?

sayakpaul replied:
Makes sense, thanks! However, EDMEuler doesn't have a `get_velocity()` function. Should we maybe tackle this is a separate PR and throw an error here instead?

sayakpaul replied:
Done in dde75950.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1591,10 +1683,23 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
                     # Compute prior loss
-                    prior_loss = F.mse_loss(model_pred_prior.float(), target_prior.float(), reduction="mean")
+                    if args.do_edm_style_training:
+                        prior_loss = torch.mean(
+                            ((model_pred_prior.float() - target_prior.float()) ** 2).reshape(
+                                target_prior.shape[0], -1
+                            ),
+                            1,
+                        )
+                        prior_loss = prior_loss.mean()
```
since we are not weighing the loss here, we could simply use `F.mse_loss`.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1591,10 +1683,23 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
                 if args.snr_gamma is None:
-                    loss = F.mse_loss(model_pred.float(), target.float(), reduction="mean")
+                    if args.do_edm_style_training:
+                        loss = torch.mean(((model_pred.float() - target.float()) ** 2).reshape(target.shape[0], -1), 1)
+                        loss = loss.mean()
```
same comment as above.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1696,8 +1801,9 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
                 )
-
                 pipeline_args = {"prompt": args.validation_prompt}
+                if args.do_edm_style_training:
+                    pipeline_args.update({"guidance_scale": 5.0})
```
Not in favour of this. value of `guidance_scale` isn't really related to `EDM`, here we should use the pipeline default or accept this as arg.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1758,6 +1864,8 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
         images = []
         if args.validation_prompt and args.num_validation_images > 0:
             pipeline_args = {"prompt": args.validation_prompt, "num_inference_steps": 25}
+            if args.do_edm_style_training:
+                pipeline_args.update({"guidance_scale": 5.0})
```
same comment as above.

patil-suraj reviewed (commented):
Looking good, left some comments.

sayakpaul commented:
@patil-suraj ready for another review.

patil-suraj commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -119,27 +136,35 @@ def save_model_card(
         repo_id_or_path=repo_id,
         from_training=True,
-        license="openrail++",
+        license="openrail++" if "playgroundai" not in base_model else "other",
```
```suggestion
        license="openrail++" if "playgroundai" not in base_model else "playground-v2dot5-community",
```

sayakpaul commented:
@pcuenca @patil-suraj I have addressed all your comments. Would appreciate another review. 

I am going to run with the command from the OP one more time and also with regular SDXL with `--do_edm_style_training`.

sayakpaul commented:
Started a regular SDXL run with EDM:

```bash
CUDA_VISIBLE_DEVICES=1 accelerate launch train_dreambooth_lora_sdxl.py \
  --pretrained_model_name_or_path="stabilityai/stable-diffusion-xl-base-1.0"  \
  --instance_data_dir="dog"\
  --pretrained_vae_model_name_or_path="madebyollin/sdxl-vae-fp16-fix" \
  --output_dir="lora-sdxl-dog" \
  --mixed_precision="fp16" \
  --use_8bit_adam \
  --do_edm_style_training \
  --instance_prompt="a photo of sks dog" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=500 \
  --validation_prompt="A photo of sks dog in a bucket" \
  --validation_epochs=25 \
  --seed="0" 
```

Garbage results: https://wandb.ai/sayakpaul/dreambooth-lora-sd-xl/runs/bup8u1yc. Let me further tweak around some things.

sayakpaul commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -1570,18 +1670,38 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
+                    # There might be other alternatives for weighting as well:
+                    # https://github.com/huggingface/diffusers/pull/7126#discussion_r1505404686
+                    if "EDM" not in scheduler_type:
+                        weighting = (sigmas**-2.0).float()
```
It is: https://wandb.ai/sayakpaul/dreambooth-lora-sd-xl/runs/dz77sffl. When `do_edm_style_training` is True and the scheduler is not EDM*, we are using EulerDiscrete. The run is from that setting.

Does that work?

patil-suraj replied:
Sounds good!

sayakpaul commented:
> Also are the vae weights loaded and kept in fp32 ?

Yes, that is the case. I have addressed your other comment as well, @patil-suraj. LMK.

patil-suraj reviewed (approved):
Looks great now! Feel free to merge

Benji-Montgomery commented:
I keep getting a PermissionError: [Errno 13] Permission denied when trying to access the dog folder. Logged in to huggingface and running as administrator. All folders have full read/write access.

sayakpaul commented:
> I keep getting a PermissionError: [Errno 13] Permission denied when trying to access the dog folder. Logged in to huggingface and running as administrator. All folders have full read/write access.

That is an issue quite unrelated to this PR.

bghira commented on examples/dreambooth/train_dreambooth_lora_sdxl.py:
```diff
@@ -905,6 +943,9 @@ def main(args):
             " Please use `huggingface-cli login` to authenticate with the Hub."
         )
 
+    if args.do_edm_style_training and args.snr_gamma is not None:
+        raise ValueError("Min-SNR formulation is not supported when conducting EDM-style training.")
```
do this earlier, so it doesn't load the model yet.

sayakpaul replied:
It's at the beginning:

https://github.com/huggingface/diffusers/blob/59433ca1ae76445908e63fe4e2f9aceaf95baf0c/examples/dreambooth/train_dreambooth_lora_sdxl.py#L948

way before the model loading code.
